File: api/label.go
package api

const (
	// EnvironmentLabelKey is the reserved key for environment.
	EnvironmentLabelKey string = "bb.environment"
	// TenantLabelKey is the label key for tenant.
	TenantLabelKey = "bb.tenant"
	// LocationLabelKey is the label key for location.
	LocationLabelKey = "bb.location"

	// DatabaseLabelSizeMax is the maximum size of database labels.
	DatabaseLabelSizeMax = 4
	labelLengthMax       = 63
)

// DatabaseLabel is the label associated with a database.
type DatabaseLabel struct {
	Key   string `json:"key"`
	Value string `json:"value"`
}

// DatabaseLabelFind finds the labels associated with the database.
type DatabaseLabelFind struct {
	// Standard fields
	RowStatus *RowStatus

	// Related fields
	DatabaseID int
}

// DatabaseLabelUpsert upserts the label associated with the database.
type DatabaseLabelUpsert struct {
	// Standard fields
	// Value is assigned from the jwt subject field passed by the client.
	UpdaterID int
	RowStatus RowStatus

	// Related fields
	DatabaseID int
	Key        string

	// Domain specific fields
	Value string
}

File: frontend/src/types/label.ts
import { LabelId } from "./id";

export type LabelKeyType = string;
export type LabelValueType = string;

export type Label = {
  id: LabelId;

  // Domain specific fields
  key: LabelKeyType;
  valueList: LabelValueType[];
};

export type LabelPatch = {
  valueList?: LabelValueType[];
};

export type DatabaseLabel = {
  key: LabelKeyType;
  value: LabelValueType;
};

export type AvailableLabel = {
  key: LabelKeyType;
  valueList: LabelValueType[];
};
